File: python/day7_part1_2025.py

import sys

def main() -> None:
    lines = sys.stdin.read().splitlines()
    if not lines:
        return

    h, w = len(lines), len(lines[0])
    grid = [list(row) for row in lines]

    # locate start
    for y in range(h):
        try:
            x = grid[y].index('S')
            sy, sx = y, x
            break
        except ValueError:
            continue
    else:
        return

    splits = 0
    active = [False] * w
    next_row = [False] * w
    active[sx] = True

    for y in range(sy, h):
        # reset next_row
        for i in range(w):
            next_row[i] = False

        row = grid[y]
        for x in range(w):
            if not active[x]:
                continue
            if row[x] == '^':
                splits += 1
                if x:
                    next_row[x - 1] = True
                if x + 1 < w:
                    next_row[x + 1] = True
            else:
                next_row[x] = True

        if not any(next_row):
            break
        active, next_row = next_row, active   # swap buffers

    print(splits)


if __name__ == "__main__":
    # read from file "input.txt" as required
    with open("input.txt", "r") as f:
        sys.stdin = f
        main()

File: lua/day7_part1_2025.lua

local function main()
    -- read whole input
    local lines = {}
    for line in io.lines("input.txt") do
        table.insert(lines, line)
    end
    local height = #lines
    if height == 0 then return end
    local width = #lines[1]

    -- locate start (S)
    local sx, sy = nil, nil
    for y = 1, height do
        local row = lines[y]
        local pos = row:find('S', 1, true)
        if pos then
            sx, sy = pos, y
            break
        end
    end
    if not sx then return end

    -- active columns for the current row
    local active = {}
    for i = 1, width do active[i] = false end
    active[sx] = true

    local splits = 0
    local next = {}

    for y = sy, height do
        -- clear next
        for i = 1, width do next[i] = false end

        local row = lines[y]
        for x = 1, width do
            if active[x] then
                local ch = row:sub(x, x)
                if ch == '^' then
                    splits = splits + 1
                    if x > 1 then next[x - 1] = true end
                    if x < width then next[x + 1] = true end
                else
                    next[x] = true
                end
            end
        end

        -- swap tables
        active, next = next, active

        -- stop if nothing is active
        local any = false
        for i = 1, width do
            if active[i] then any = true; break end
        end
        if not any then break end
    end

    print(splits)
end

main()
